File: src/org/odk/collect/android/activities/GeoPointActivity.java
/*
 * Copyright (C) 2009 University of Washington
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package org.odk.collect.android.activities;

import java.text.DecimalFormat;
import java.util.List;

import org.odk.collect.android.R;
import org.odk.collect.android.application.GeoProgressDialog;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.location.LocationProvider;
import android.os.Bundle;
import android.view.View;
import android.view.View.OnClickListener;

public class GeoPointActivity extends Activity implements LocationListener {
    private GeoProgressDialog mLocationDialog;
    private LocationManager mLocationManager;
    private Location mLocation;
    private boolean mGPSOn = false;
    private boolean mNetworkOn = false;
    
    private int acceptableThreshold = 1600;

    // default location accuracy
    private static double LOCATION_ACCURACY = 5;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setTitle(getString(R.string.app_name) + " > " + getString(R.string.get_location));

        mLocationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
        
        evalProviders();
        
    	setupLocationDialog();

    }


	private void evalProviders() {

        // make sure we have a good location provider before continuing
        List<String> providers = mLocationManager.getProviders(true);
        for (String provider : providers) {
            if (provider.equalsIgnoreCase(LocationManager.GPS_PROVIDER)) {
                mGPSOn = true;
            }
            if (provider.equalsIgnoreCase(LocationManager.NETWORK_PROVIDER)) {
                mNetworkOn = true;
            }
        }

	}


	private void showNoGpsDialog() {
		AlertDialog dialog = new AlertDialog.Builder(this).create();
		dialog.setTitle(getString(R.string.no_gps_title));
		dialog.setMessage(getString(R.string.no_gps_message));
        DialogInterface.OnClickListener changeSettingsListener = new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int i) {
                switch (i) {
                    case DialogInterface.BUTTON1: //Yes, get settings 
                		Intent intent = new Intent(android.provider.Settings.ACTION_LOCATION_SOURCE_SETTINGS);
                		startActivity(intent);
                        break;
                    case DialogInterface.BUTTON2: //No, bail
                    	mLocation = null;
                    	GeoPointActivity.this.finish();
                    	break;
                }
            }
        };
        

        DialogInterface.OnCancelListener onCancelListener = new DialogInterface.OnCancelListener() {
			@Override
			public void onCancel(DialogInterface dialog) {
            	mLocation = null;
            	GeoPointActivity.this.finish();
			}
        };
        
        dialog.setCancelable(true);
        dialog.setOnCancelListener(onCancelListener);
        dialog.setButton(getString(R.string.change_settings), changeSettingsListener);
        dialog.setButton2(getString(R.string.cancel), changeSettingsListener);

        dialog.show();
	}


	@Override
    protected void onPause() {
        super.onPause();

        // stops the GPS. Note that this will turn off the GPS if the screen goes to sleep.
        mLocationManager.removeUpdates(this);

        // We're not using managed dialogs, so we have to dismiss the dialog to prevent it from
        // leaking memory.
        if (mLocationDialog != null && mLocationDialog.isShowing())
            mLocationDialog.dismiss();
    }


    @Override
    protected void onResume() {
        super.onResume();
        evalProviders();
        if (!mGPSOn && !mNetworkOn) {
            showNoGpsDialog();
        } else {
	        if (mGPSOn) {
	            mLocationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, this);            
	        }
	        if (mNetworkOn) {
	            mLocationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 0, 0, this);
	        }
	        mLocationDialog.show();
        }
    }


    /**
     * Sets up the look and actions for the progress dialog while the GPS is searching.
     */
    private void setupLocationDialog() {
        // dialog displayed while fetching gps location
        
        OnClickListener cancelButtonListener =
            new OnClickListener() {
                @Override
    	   		public void onClick(View v){
                    mLocation = null;
                    finish();
    	   		}
        	};
        	
        OnClickListener okButtonListener =
    		new OnClickListener() {
    	   		public void onClick(View v){
                    returnLocation();
    	   		}
       	};
       	
       	mLocationDialog = new GeoProgressDialog(this, getString(R.string.finding_location), getString(R.string.found_location));
       	

        // back button doesn't cancel
        mLocationDialog.setCancelable(false);
        mLocationDialog.setImage(getResources().getDrawable(R.drawable.green_check_mark));
        mLocationDialog.setMessage(getString(R.string.please_wait_long));
        mLocationDialog.setOKButton(getString(R.string.accept_location),
            okButtonListener);
        mLocationDialog.setCancelButton(getString(R.string.cancel_location),
            cancelButtonListener);
    }


    private void returnLocation() {
        if (mLocation != null) {
            Intent i = new Intent();
            i.putExtra(
                FormEntryActivity.LOCATION_RESULT,
                mLocation.getLatitude() + " " + mLocation.getLongitude() + " "
                        + mLocation.getAltitude() + " " + mLocation.getAccuracy());
            setResult(RESULT_OK, i);
        }
        finish();
    }


    @Override
    public void onLocationChanged(Location location) {
        mLocation = location;
        if (mLocation != null) {
            mLocationDialog.setMessage(getString(R.string.location_provider_accuracy,
                mLocation.getProvider(), truncateDouble(mLocation.getAccuracy())));

            if (mLocation.getAccuracy() <= LOCATION_ACCURACY) {
                returnLocation();
            }
            
            if(mLocation.getAccuracy() < acceptableThreshold){
            	mLocationDialog.setLocationFound(true);
            } else{
            	mLocationDialog.setLocationFound(false);
            }
        }
    }


    private String truncateDouble(float number) {
        DecimalFormat df = new DecimalFormat("#.##");
        return df.format(number);
    }


    @Override
    public void onProviderDisabled(String provider) {

    }


    @Override
    public void onProviderEnabled(String provider) {

    }


    @Override
    public void onStatusChanged(String provider, int status, Bundle extras) {
        switch (status) {
            case LocationProvider.AVAILABLE:
                if (mLocation != null) {
                    mLocationDialog.setMessage(getString(R.string.location_accuracy,
                        mLocation.getAccuracy()));
                }
                break;
            case LocationProvider.OUT_OF_SERVICE:
                break;
            case LocationProvider.TEMPORARILY_UNAVAILABLE:
                break;
        }
    }

}

File: src/org/odk/collect/android/activities/GeoPointActivity.scala
package org.odk.collect.android.activities

import android.location.LocationListener
import android.os.Bundle
import android.app.Activity
import org.odk.collect.android.listeners.TimerListener
import org.odk.collect.android.application.GeoProgressDialog
import android.location.Location
import android.location.LocationManager
import org.odk.collect.android.utilities.ODKTimer
import org.odk.collect.android.R
import org.odk.collect.android.utilities.GeoUtils
import android.content.Context
import java.util.Set
import android.content.DialogInterface
import scala.collection.JavaConversions._
import android.content.Intent
import android.view.View.OnClickListener
import android.view.View
import java.text.DecimalFormat
import android.location.LocationProvider

class GeoPointActivity extends Activity with LocationListener with TimerListener {
  var mLocationDialog: GeoProgressDialog = null
  var mLocationManager: LocationManager = null
  var mLocation: Location = null
  var mProviders: java.util.Set[String] = null
  var mTimer: ODKTimer = null
  
  val millisToWait: Int = 60000 // allow to accept location after 60 seconds
  
  override def onCreate(savedInstanceState: Bundle) {
    super.onCreate(savedInstanceState)
    setTitle(getString(R.string.app_name) + " > " + getString(R.string.get_location))
    mLocationManager = getSystemService(Context.LOCATION_SERVICE).asInstanceOf[LocationManager]
    mProviders = GeoUtils.evaluateProviders(mLocationManager)
    
    setupLocationDialog()

    var mLong: Long = -1
    if (savedInstanceState != null) {
        mLong = savedInstanceState.getLong("millisRemaining",-1)
    }
    if (mLong > 0) {
        mTimer = new ODKTimer(mLong, this)
    }
    else {
        mTimer = new ODKTimer(millisToWait, this)
    }
        
    mTimer.start()
  }
  
  override def onPause() {
    super.onPause()
    
    // Stop the GPS. Note that this will turn off the GPS if the screen goes to sleep.
    mLocationManager.removeUpdates(this)

    // We're not using managed dialogs, so we have to dismiss the dialog to prevent it from leaking memory.
    if (mLocationDialog != null && mLocationDialog.isShowing()) mLocationDialog.dismiss()
  }
  
  override def onResume() {
    super.onResume()
    
    mProviders = GeoUtils.evaluateProviders(mLocationManager)
    
    if (mProviders.isEmpty()) {
      var onCancelListener: DialogInterface.OnCancelListener = new DialogInterface.OnCancelListener() {
        override def onCancel(dialog: DialogInterface) {
            mLocation = null
            GeoPointActivity.this.finish()
        }
      }
      
      var onChangeListener: DialogInterface.OnClickListener = new DialogInterface.OnClickListener() {
        override def onClick(dialog: DialogInterface, i: Int) {
          i match {
            case DialogInterface.BUTTON_POSITIVE => {
              var intent: Intent = new Intent(android.provider.Settings.ACTION_LOCATION_SOURCE_SETTINGS)
              startActivity(intent)
            }
            case DialogInterface.BUTTON_NEGATIVE => {
              mLocation = null
              GeoPointActivity.this.finish()
            }
          }
        }
      }
      
      GeoUtils.showNoGpsDialog(this, onChangeListener, onCancelListener)
    }
    else {
      for (provider <- mProviders) {
        mLocationManager.requestLocationUpdates(provider, 0, 0, this)
      }
      mLocationDialog.show()
    }
  }
  
  def setupLocationDialog() {
    var cancelButtonListener: OnClickListener = new OnClickListener() {
      override def onClick(v: View) {
        mLocation = null
        finish()
      }
    }
    
    var okButtonListener: OnClickListener = new OnClickListener() {
      override def onClick(v: View) {
        returnLocation()
      }
    }
    
    mLocationDialog = new GeoProgressDialog(this, getString(R.string.found_location), getString(R.string.finding_location))

    mLocationDialog.setCancelable(false)
    mLocationDialog.setImage(getResources().getDrawable(R.drawable.green_check_mark))
    mLocationDialog.setMessage(getString(R.string.please_wait_long))
    mLocationDialog.setOKButton(getString(R.string.accept_location), okButtonListener)
    mLocationDialog.setCancelButton(getString(R.string.cancel_location), cancelButtonListener)
  }
  
  def returnLocation() {
    if (mLocation != null) {
      var i: Intent = new Intent()
      i.putExtra(FormEntryActivity.LOCATION_RESULT, GeoUtils.locationToString(mLocation))
      setResult(Activity.RESULT_OK, i)
    }
    finish()
  }
  
  override def onLocationChanged(location: Location) {
    mLocation = location
    if (mLocation != null) {
      mLocationDialog.setMessage(getString(R.string.location_provider_accuracy, mLocation.getProvider(), truncateDouble(mLocation.getAccuracy())))

      // If location is accurate, we're done
      if (mLocation.getAccuracy() <= GeoUtils.GOOD_ACCURACY) {
        returnLocation()
      }

      // If location isn't great but might be acceptable, notify
      // the user and let them decide whether or not to record it
      mLocationDialog.setLocationFound(mLocation.getAccuracy() < GeoUtils.ACCEPTABLE_ACCURACY || mTimer.getMillisUntilFinished() == 0)
    }
  }
  
  def truncateDouble(number: Float): String = {
    var df: DecimalFormat = new DecimalFormat("#.##")
    df.format(number)
  }
  
  override def onProviderDisabled(provider: String) {}

  override def onProviderEnabled(provider: String) {}
            
  override def onStatusChanged(provider: String, status: Int, extras: Bundle) {
    if (status == LocationProvider.AVAILABLE) {
      if (mLocation != null) {
        mLocationDialog.setMessage(getString(R.string.location_accuracy, new Integer(mLocation.getAccuracy().toInt)))
      }
    }
  }
  
  override def notifyTimerFinished() {
    onLocationChanged(mLocation)
  }
  
  override def onSaveInstanceState(savedInstanceState: Bundle) {
    savedInstanceState.putLong("millisRemaining", mTimer.getMillisUntilFinished())
    super.onSaveInstanceState(savedInstanceState)
  }

}